proxy/src/control/discovery.rs

```diff
@@ -154,44 +188,106 @@ impl Discovery {
 
-impl<B> Discover for Watch<B>
+impl<B, A> Discover for Watch<B>
 where
-    B: Bind,
+    B: Bind<Request = http::Request<A>>,
 {
     type Key = SocketAddr;
     type Request = B::Request;
     type Response = B::Response;
     type Error = B::Error;
-    type Service = B::Service;
+    type Service = Labeled<B::Service>;
     type DiscoverError = ();
 
     fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::DiscoverError> {
-        let up = self.rx.poll();
-        trace!("watch: {:?}", up);
-        let update = match up {
-            Ok(Async::Ready(Some(update))) => update,
-            Ok(Async::Ready(None)) => unreachable!(),
-            Ok(Async::NotReady) => return Ok(Async::NotReady),
-            Err(_) => return Err(()),
-        };
-
-        match update {
-            Update::Insert(addr) => {
-                let service = self.bind.bind(&addr).map_err(|_| ())?;
+        loop {
+            let up = self.rx.poll();
+            trace!("watch: {:?}", up);
+            let update = match up {
+                Ok(Async::Ready(Some(update))) => update,
+                Ok(Async::Ready(None)) => unreachable!(),
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Err(_) => return Err(()),
+            };
 
-                Ok(Async::Ready(Change::Insert(addr, service)))
-            },
-            // TODO: handle metadata changes by changing the labeling
-            // middleware to hold a `futures-watch::Watch` on the label value,
-            // so it can be updated.
-            Update::Remove(addr) => Ok(Async::Ready(Change::Remove(addr))),
+            match update {
+                Update::Insert(addr, meta) => {
+                    // Construct a watch for the `Labeled` middleware that will
+                    // wrap the bound service, and insert the store into our map
+                    // so it can be updated later.
+                    let (watch_labels, update_labels) =
+                        futures_watch::Watch::new(meta.metric_labels);
+                    self.metric_labels.insert(addr, update_labels);
+
+                    let service = self.bind.bind(&addr)
+                        .map(|svc| Labeled::new(svc, watch_labels))
+                        .map_err(|_| ())?;
+
+                    return Ok(Async::Ready(Change::Insert(addr, service)))
+                },
+                Update::ChangeMetadata(addr, meta) => {
+                    // Update metadata and continue polling `rx`.
+                    self.update_metadata(addr, meta)?;
+                },
+                Update::Remove(addr) => {
+                    // NOTE: we don't remove the `store` for the removed
+                    // service's labels from `metrics_labels` here, since the
+                    // balancer may not have removed it yet. If we see another
+                    // ChangeMetadata event for that watch and it's cancelled,
+                    // we'll remove it then.
```

Comment:
eh, sorry. on further reflection, my suggestion wasn't right. (though, your comment here is a huge help to realize that).

this is a memory leak.  We need to remove the item from metric_labels here.

before doing that, we need to confirm that the Watch still functions properly after the Store is dropped. i'm 99% sure it does, but we should check that.

Reply:
It looks like `futures-watch` has a test for that : https://github.com/carllerche/better-future/blob/07baa13e91fefe7a51533dfde7b4e69e109ebe14/futures-watch/tests/watch.rs#L30-L41